Cleanup code

kerim1 opened the pull request:
I looked into the changes you made to the codebase and I cleaned them up according to how this library was set up. It should still work with your use case, can you check?

1. Properties are updated through various methods; preferably not directly. I cannot mark them private as it then won't show up in the JSON structure though. :-) 
2. If a structure can be used in various types, the code is found in the `Base` class. This prevents a lot of code duplication. For example: a lot of structures have `items`, so there is a method `setItems()` in the `Base` class.
3. Don't set the type directly: this is taken care of in the constructor already! 
4. Same for the context: there is a method `setContext()` for this purpose. In some cases you don't want a context and in other cases you want to add another couple of context JSON URLs.
5. I found some incorrect properties. For example, you added a property `start` to `Canvas`. But according to the documentation, that is not allowed!

Sozialarchiv commented:
I try to find a compromis between your suggestions and my modifications:

> 1. Properties are updated through various methods; preferably not directly. I cannot mark them private as it then won't show up in the JSON structure though. :-)

Fixed: All properties should now have methods.

> 2. If a structure can be used in various types, the code is found in the Base class. This prevents a lot of code duplication. For example: a lot of structures have items, so there is a method setItems() in the Base class.

The common setItems functions caused issues: https://github.com/archival-IIIF/presentation-builder/issues/1. 

Of course, the approach to prevent duplicated code is good. However, in this case I think it is acceptable for a little duplicated code to be specification compliant. E.g. Canvas has no viewingDirection.

> 3. Don't set the type directly: this is taken care of in the constructor already!

I just did that in classes (Provider and Image) which do not extends Base

> 4. Same for the context: there is a method setContext() for this purpose. In some cases you don't want a context and in other cases you want to add another couple of context JSON URLs.

Fixed

> 5. I found some incorrect properties. For example, you added a property start to Canvas. But according to the documentation, that is not allowed!

Fixed: If you see other things please write.

I removed the Rendering class in v3 like you suggested.

I checked the server repository. The code is working with this compromise. See https://github.com/archival-IIIF/server/tree/presentation-builder-npm

Is this code for you acceptable?

kerim1 commented:
Sorry, had little time to work on this, but here I am! :-)

I cleaned up the code again from the latest version in the master branch. My feedback:

> If a structure can be used in various types, the code is found in the Base class. This prevents a lot of code duplication. For example: a lot of structures have items, so there is a method setItems() in the Base class.

> The common setItems functions caused issues: #1.
Of course, the approach to prevent duplicated code is good. However, in this case I think it is acceptable for a little duplicated code to be specification compliant. E.g. Canvas has no viewingDirection.

That is the hard part about translating the spec to code, various properties are allowed in various types, but not all types. I see the compromises you made. I have created a helper function that deals with the setter code, to avoid a lot of the duplication, but still allow separate `setItem` functions for example.

The issue you refer to is because I created the library for writing manifests; not for reading them. However, now with separate `setItem` functions that should work as well.

> Don't set the type directly: this is taken care of in the constructor already!

> I just did that in classes (Provider and Image) which do not extends Base

Yet I still encountered them and I removed them! :-) 

> Same for the context: there is a method setContext() for this purpose. In some cases you don't want a context and in other cases you want to add another couple of context JSON URLs.

> Fixed

Again I encountered them: I fixed them and fixed the unit test: they were not calling the method `setContext()` and were missing the W3 annotations URI.

Please accept the pull request or request changes here!

And when this is finished, I can start adapting the server code to use the NPM library instead. However: if I want to make changes like adding another property or something, how can I push a new version to NPM?
